### PR TITLE
Link to types in Version module

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

**Comment by fertapric** [comment]
It seems types are directly linked in the documentation using `t:my_type/0` in all cases except for [this case](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/enum.ex#L2138). This would introduce a few more.

I'm not sure about the new links 😕. For example, the link ``` [string](`t:String.t/0`) ``` does not seem really helpful, and it would be insane to link all the mentions to "string" in the documentation. On top of that, the readability of the documentation from IEx or the source code gets affected:

```
A [version](`t:version/0`) is a [string](`t:String.t/0`) in a specific
format or a [`%Version{}`](`t:Version.t/0`) struct
generated after parsing a version string with `Version.parse/1`.
```

V.S.

```
A version is a string in a specific format or a `Version` 
generated after parsing via `Version.parse/1`.
```

Also, I would use the module name instead of ``` [`%Version{}`](`t:Version.t/0`) ```. The module documentation usually contains a more detailed documentation than the type; [`Version`](https://hexdocs.pm/elixir/Version.html#content) and [`Version.t()`](https://hexdocs.pm/elixir/Version.html#t:t/0) are actually good examples of this. Besides, the argument related to the readability of the documentation from IEx or the source code can also be applied in this case. 

I like the link to SemVer though 🙂

**Comment by eksperimental** [comment]
We can definitely remove the link to the string type.
If IEx is misformatting this, then it is something that we need to fix there.

The reason we we never use this format elsewhere could be due to the fact that ExDoc was not supporting custom links for some types, for example built-in types. And that is something it is getting fixed in ExDoc.

**Comment by fertapric** [comment]
IEx formats the links using parentheses. So it would look like (with some coloring):

> A  version (`t:version/0`) is a string (`t:String.t/0`) in a specific format or a %Version{} (`t:Version.t/0`) struct generated after parsing a version string with `Version.parse/1`.

compared to:

> A  version is a string in a specific format or a `Version` struct generated after parsing a version string with `Version.parse/1`.

**Comment by eksperimental** [comment]
>> Also, I would use the module name instead of [`%Version{}`](`t:Version.t/0`). The module documentation usually contains a more detailed documentation than the type; Version and Version.t() are actually good examples of this.

Sorry, but why would you like to the module itself in the first paragraph of module documentation?

**Comment by fertapric** [comment]
> Sorry, but why would you like to the module itself in the first paragraph of module documentation?

Ah! Sorry. I thought this was linking to an external module. Sorry for the noise.

**Comment by eksperimental** [comment]
once we remove the link to the string type, and linking to `%Version{} struct`, 
will look something like this in IEx (which is is perfectly fine to me):

> A version (t:version/0) is a string in a specific format or a %Version{} struct (t:Version.t/0) generated after parsing a version string with Version.parse/1.

what you do you think?

**Comment by fertapric** [comment]
> what you do you think?

👍

File: lib/elixir/lib/version.ex [hunk 1]
```diff
@@ -24,8 +25,8 @@ defmodule Version do
 
   ## Struct
 
-  The version is represented by the `Version` struct and fields
-  are named according to SemVer: `:major`, `:minor`, `:patch`,
+  The version is represented by the [`%Version{}`](`t:Version.t/0`) struct and fields
+  are named according to [SemVer](http://semver.org/): `:major`, `:minor`, `:patch`,
```
Comment (ericmj) [inline lib/elixir/lib/version.ex]: We don't have to link to the semver website multiple times, we already have a link on line 10.

**Comment by josevalim** [comment]
I agree with @fertapric. We should avoid introducing link to types in the documentation, that's what typespecs are for. It is preferrable to say "a `Range`" rather than link to the range type. Or say "a `Version` struct".

**Pull request closed without merging**